pkg/boot/fit/vfit.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2021 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+//
+// Performs signature checks on FDT images.
+// Currently supports PGP and raw PKCS1v15 RSA signatures.
+//
+// Expected FDT Format:
+//  Node: images
+//  Node: image_name
+//   P: data
+//   Node: signature*
+//    P: value
+//    P: algo          (ex. 'sha256,rsa4096', 'pgp')
+//    P: signer-name   (Optional)
+//    P: key-name-hint (Optional)
+//
+// vfit aims to be TOCTTOU-safe by reading files into memory before verifying.
+package fit
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rsa"
+	"fmt"
+	"strings"
+	"unicode"
+
+	"github.com/u-root/u-root/pkg/dt"
+	"github.com/u-root/u-root/pkg/vfile"
+	"golang.org/x/crypto/openpgp"
+)
+
+var (
+	algs = map[string]crypto.Hash{
+		"MD4":       crypto.MD4,
+		"MD5":       crypto.MD5,
+		"SHA1":      crypto.SHA1,
+		"SHA224":    crypto.SHA224,
+		"SHA256":    crypto.SHA256,
+		"SHA384":    crypto.SHA384,
+		"SHA512":    crypto.SHA512,
+		"RIPEMD160": crypto.RIPEMD160,
+		"SHA3_224":  crypto.SHA3_224,
+		"SHA3_256":  crypto.SHA3_256,
+		"SHA3_384":  crypto.SHA3_384,
+		"SHA3_512":  crypto.SHA3_512,
+	}
+)
+
+type Signature interface {
+	fmt.Stringer
+	// Warning: If the signature does not exist or does not match the keyring,
+	// both the file and a signature error will be returned.
+	Verify([]byte, openpgp.KeyRing) (*bytes.Reader, error)
```

Comment:
Why make this function return a reader?

Reply:
Largely just for TOCTTOU-safety. That way we make it harder for the image that was checked to be altered. I followed the standard of vfile here.

Reply:
Ah, I see. Can you add a comment about this here?

Also optional: You could return a [io.ReadSeeker](https://pkg.go.dev/io#ReadSeeker) so that the underlying implementation is free to change (i.e., use something different than `bytes.Reader`). This is completely optional since it doesn't seem to have been needed for vfile.

Reply:
I agree it's a little moot since it's pointing back to their input data where currently used vfile is largely taking a path..
Aside from duplicating large kernels/initrams in memory which I'm reluctant to do in this scenario, I think it's fine.
Squahed a brief comment change for the return.